### DISCUSSION TITLE
How can I pass a build setting to a genrule?

### DISCUSSION THREAD

**Discussion by njlr** [bazelbuild/bazel#18600] [Q&A]
Suppose I have a simple `genrule` like this:

```starlark
genrule(
  name = "fruit",
  outs = [ "fruit.txt" ],
  cmd = "echo \"fruit: apple\" > $@",
)
```

Now I want to make this configurable from the outside.

I can use `--action_env` like this:

```starlark
genrule(
  name = "fruit",
  outs = [ "fruit.txt" ],
  cmd = "echo \"fruit: $$FRUIT\" > $@",
)
```

```bash
bazel build --action_env=FRUIT=banana //:fruit
```

However, I don't want to use `--action_env` because it affects all rules and it is bad for caching.

There is [`--define`](https://bazel.build/reference/command-line-reference#flag--define), but the docs describe it as legacy:

> `--define` is an alternative legacy syntax for custom flags (for example `--define foo=bar`). This can be expressed either in the values attribute (`values = {"define": "foo=bar"}`) or the define_values attribute (`define_values = {"foo": "bar"}`). `--define` is only supported for backwards compatibility. Prefer Starlark build settings whenever possible.

So that leaves build settings.

However, it is not clear how to get the value of a build setting into my `genrule`.

I am looking for something like this:

```starlark
# Not real code 

config_setting(
  name = "favourite_fruit",
)

genrule(
  name = "fruit",
  outs = [ "fruit.txt" ],
  cmd = "echo \"fruit: $(value :favourite_fruit)\" > $@",
)
```

**Comment by meteorcloudy** [comment]
Maybe examples under https://github.com/bazelbuild/examples/tree/main/configurations can help?

/cc @gregestren @katre

**Reply by njlr** [reply]
> Maybe examples under https://github.com/bazelbuild/examples/tree/main/configurations can help?
> 
> /cc @gregestren @katre

Thanks for the links. 

Those appear to require implementing a custom rule. Is it possible to simply use `genrule` so as to avoid this?

**Comment by gregestren** [comment]
Hi @njlr,

Build settings work best with `select()` and with custom rules that are designed to know how to read them. You have a case where `--define`  is more powerful: build settings don't automatically work with [make variable substitution](https://bazel.build/reference/be/make-variables) like `--define` does.

I can think of 3 approaches:

1. Just use --`define` in this case, in spite of the legacy documentation
2. Write [some Starlark](https://github.com/bazelbuild/examples/blob/6f10259925ccb8f22875af4e1a88c0872ee7dccc/make-variables/testapp/defs.bzl) that tells Bazel how to connect your build setting into a `cmd = $(foo) ` variable, then register that setting in your `genrule`'s [toolchains](https://github.com/bazelbuild/examples/blob/6f10259925ccb8f22875af4e1a88c0872ee7dccc/make-variables/testapp/BUILD#L89) attribute (also see [here](https://bazel.build/reference/be/make-variables#custom_variables))
3. Have your genrule use a `select()` like [this](https://gist.github.com/gregestren/e7afc5a3c3622af326ec49c9e2419c3e#file-gistfile1-txt)

Each approach has its pluses and minuses. If you just want something quick and easy and don't want to have to add any extra boilerplate, I'd just stick with `--define`.

Otherwise I'd use option 3 if it works, with the biggest weakness that you need a `config_setting` for every possible value the flag might take.

Option 2 should work for any value but has the most boilerplate.
